ir/memory.cpp

```diff
@@ -18,15 +18,67 @@ using namespace util;
 
   // Non-local block ids (assuming that no block is optimized out):
   // 1. null block: 0
-  // 2. global vars in source: 1 ~ num_globals_src
+  // 2. global vars in source:
+  //      1 ~ num_consts_src                    (constant globals)
+  //      num_consts_src + 1 ~ num_globals_src  (non-constant globals)
   // 3. pointer argument inputs:
   //      num_globals_src + 1 ~ num_globals_src + num_ptrinputs
   // 4. a block reserved for encoding the memory touched by calls:
   //      num_globals_src + num_ptrinputs + 1
   // 5. nonlocal blocks returned by loads/calls:
   //      num_globals_src + num_ptrinputs + 2 ~ num_nonlocals_src - 1:
   // 6. global vars in target only:
-  //      num_nonlocals_src ~ num_nonlocals - 1
+  //      num_nonlocals_src ~ num_nonlocals_src + num_extra_nonconst_tgt - 1
+  //            (non-constant globals in target only)
+  //      num_nonlocals_src + num_extra_nonconst_tgt ~ num_nonlocals - 1
+  //            (constant globals in target only)
+
+//--- Functions for non-local block analysis based on bid ---//
+
+// If include_tgt is true, return true if bid is a global var existing in target
+// only as well
+static bool nonlocal_is_globalvar(unsigned bid, bool include_tgt) {
+  bool srcglb = has_null_block <= bid && bid < has_null_block + num_globals_src;
```

Comment:
the first bit is redundant

Reply:
I think it is good to include non-null globals only.
It allows us to reason that bid s.t. is_globalvar(bid) is alive.

Reply:
it seems global vars are assumed to be live ones: line 655 in this file had a relevant comment
```
  // globals are always live
  if (local || (bid0 >= num_globals_src && bid0 < num_nonlocals_src)) {
```